Terminal.Gui/Views/SplitContainer.cs

```diff
@@ -0,0 +1,554 @@
+
+			var idealPosition = pos.Anchor (availableSpace);
+
+			// bad position because not enough space for splitterPanels[0]
+			if (idealPosition < splitterPanels [0].MinSize.Anchor (availableSpace)) {
+
+				// TODO: we should preserve Absolute/Percent status here not just force it to absolute
+				return (Pos)Math.Min (splitterPanels [0].MinSize.Anchor (availableSpace), availableSpace);
+			}
+
+			// bad position because not enough space for splitterPanels[1]
+			if (availableSpace - idealPosition <= splitterPanels [1].MinSize.Anchor (availableSpace)) {
+
+				// TODO: we should preserve Absolute/Percent status here not just force it to absolute
+
+				// +1 is to allow space for the splitter
+				return (Pos)Math.Max (availableSpace - (splitterPanels [1].MinSize.Anchor (availableSpace) + 1), 0);
+			}
+
+			// this splitter position is fine, there is enough space for everyone
+			return pos;
+		}
+
+		/// <summary>
+		/// A panel within a <see cref="SplitterPanel"/>. 
+		/// </summary>
+		public class SplitterPanel : View {
+			Pos minSize = 1;
+
+			/// <summary>
+			/// Gets or sets the minimum size for the panel.
+			/// </summary>
+			public Pos MinSize { get => minSize;
+				set { 
+					minSize = value;
+					SuperView?.SetNeedsLayout ();
+				} 
+			}
+
+			ustring title = ustring.Empty;
+			/// <summary>
+			/// The title to be displayed for this <see cref="SplitterPanel"/>. The title will be rendered 
+			/// on the top border aligned to the left of the panel.
+			/// </summary>
+			/// <value>The title.</value>
+			public ustring Title {
+				get => title;
+				set {
+					title = value;
+					SetNeedsDisplay ();
+				}
+			}
+
+			/// <inheritdoc/>
+			public override void Redraw (Rect bounds)
+			{
+				Driver.SetAttribute (ColorScheme.Normal);
+				base.Redraw (bounds);
+			}
+
+			/// <inheritdoc/>
+			public override void OnVisibleChanged ()
+			{
+				base.OnVisibleChanged ();
+				SuperView?.SetNeedsLayout ();
+			}
+		}
+
+		private class SplitContainerLineView : LineView {
+			private SplitContainer parent;
+
+			Point? dragPosition;
+			Pos dragOrignalPos;
+			Point? moveRuneRenderLocation;
+
+			public SplitContainerLineView (SplitContainer parent)
+			{
+				CanFocus = true;
+				TabStop = true;
+
+				this.parent = parent;
+
+				base.AddCommand (Command.Right, () => {
+					return MoveSplitter (1, 0);
+				});
+
+				base.AddCommand (Command.Left, () => {
+					return MoveSplitter (-1, 0);
+				});
+
+				base.AddCommand (Command.LineUp, () => {
+					return MoveSplitter (0, -1);
+				});
+
+				base.AddCommand (Command.LineDown, () => {
+					return MoveSplitter (0, 1);
+				});
+
+				AddKeyBinding (Key.CursorRight, Command.Right);
+				AddKeyBinding (Key.CursorLeft, Command.Left);
+				AddKeyBinding (Key.CursorUp, Command.LineUp);
+				AddKeyBinding (Key.CursorDown, Command.LineDown);
+
+				LayoutStarted += (e) => {
+					moveRuneRenderLocation = null;
+					if (Orientation == Orientation.Horizontal) {
```

Comment:
This should all be refactored. 

We should add to `ConsoleDriver` a new method, `DrawWindowLine`. `ConsoleDriver.DrawWindowFrame` should be refactored to use `DrawWindowLine`. `DrawWindowLine` should have a parameter `StartingAnchor` and `EndingAnchor`. 

`LineView` should also be refactored to use `DrawWindowLine` (basically we should move the `Redraw` code from `LineView` to `ConsoleDriver.DrawWindowLine`.

Reply:
I just had a crazy idea...

What if we just made it so an extended version of `DrawWindowFrame` was able to draw arbitrary nested/tiled panels?

There's got to be a way to describe a frame with infinitely nested sub-panels via some sort of data structure (I too dum to know, but I think I remember MS doing something like this back in Windows 1.x). Then all the logic trying to deal with z-order issues in `Window` and `FrameView` (and now `SplitContainer`) could be moved down to `DrawWindowFrame`?

Reply:
Almost like WM_NCPAINT of old (fun fact, I was one of the first Windows developers outside of MS to leverage WM_NCPAINT for fun and profit back in the late '80s). One of my most popular pieces of code that got shared everywhere was named "Itsybits" and enabled windows with tiny title bars.

https://learn.microsoft.com/en-us/windows/win32/gdi/nonclient-area

Just found this:

https://view.officeapps.live.com/op/view.aspx?src=https%3A%2F%2Fraw.githubusercontent.com%2FPaolo-Maffei%2FOpenNT%2F5c5b979ec08c17d3ca2eb70e8aad62d26515d01c%2Fsdktools%2Ftopdesk%2Fapp%2Fitsybits.doc&wdOrigin=BROWSELINK

LMAO!

Reply:
Code never dies! Thats so awesome.

Reply:
Also in GitHub https://gist.github.com/tig/3491637.

Reply:
Now refactored to use `LineCanvas` so resolving conversation.